### PR TITLE
HARP-11102: Support MapView's max bounds for globe.

### PR DESCRIPTION
**Pull request by atomicsulfate**
Signed-off-by: Andres Mandado <andres.mandado-almajano@here.com>

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-mapview/lib/Utils.ts [hunk 1]
```diff
@@ -377,24 +404,90 @@ export namespace MapViewUtils {
...
+                    .setZ(unconstrained.target.z);
+                if (constrained.target.equals(unconstrained.target)) {
+                    return unconstrained;
+                }
+
+                camera.position.x += constrained.target.x - unconstrained.target.x;
+                camera.position.y += constrained.target.y - unconstrained.target.y;
+            }
+            return constrained;
+        }
```
Comment (atomicsulfate) [inline @here/harp-mapview/lib/Utils.ts:429]: Planar implementation didn't change, just moved.

File: @here/harp-mapview/lib/Utils.ts [hunk 2]
```diff
@@ -162,32 +179,42 @@ export namespace MapViewUtils {
...
 
-        if (mapView.geoMaxBounds && worldTarget) {
-            const { distance } = constrainTargetAndDistanceToViewBounds(
-                worldTarget,
+        if (mapView.geoMaxBounds) {
+            // If map view has maximum bounds set, constrain camera target and distance to ensure
+            // they remain within bounds.
+
+            // tslint:disable-next-line: deprecation
+            const oldTarget = getTargetAndDistance(projection, camera, elevationProvider).target;
```
Comment (ninok) [inline @here/harp-mapview/lib/Utils.ts:187]: is this different than the current `MapView.target` projected to world space?

File: @here/harp-mapview/lib/Utils.ts [hunk 3]
```diff
@@ -162,32 +179,42 @@ export namespace MapViewUtils {
...
+            const newCamPos = setWorldPosHeight(
```
Comment (ninok) [inline @here/harp-mapview/lib/Utils.ts:188]: I have problems understanding all that code. Why do we use the camera height at all and then compute the distance? Can we not directly compute the desired distance and constrain it?

File: @here/harp-mapview/lib/Utils.ts [hunk 4]
```diff
@@ -377,24 +404,90 @@ export namespace MapViewUtils {
...
+
+                camera.position.x += constrained.target.x - unconstrained.target.x;
+                camera.position.y += constrained.target.y - unconstrained.target.y;
+            }
+            return constrained;
+        }
+
+        // Spherical projection
         if (scale > 1) {
+            // Set target to center of max bounds but keeping same height as unconstrained target.
```
Comment (ninok) [inline @here/harp-mapview/lib/Utils.ts:433]: Isn't that height always the earth radius or do we allow the target to "fly"? Would make things easier to understand if you use `EarthConstants.EQUATORIAL_RADIUS` here

File: @here/harp-mapview/lib/Utils.ts [hunk 5]
```diff
@@ -377,24 +404,90 @@ export namespace MapViewUtils {
             distance: unconstrained.distance
         };
 
+        if (projection.type === ProjectionType.Planar) {
+            if (scale > 1) {
```
Comment (ninok) [inline @here/harp-mapview/lib/Utils.ts:408]: I know it's not your code but can you please add a comment why we have to distinguish between `scale <= 1` and `scale > 1`

File: @here/harp-mapview/lib/Utils.ts (hunk 4 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/Utils.ts:433]: I wanted to make it a little more flexible to support 3D terrain.

Comment (ninok) [reply @here/harp-mapview/lib/Utils.ts:433]: okay, makes sense.

File: @here/harp-mapview/lib/Utils.ts (hunk 5 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/Utils.ts:408]: it's actually my code, merged a couple of weeks ago. I'll add the comment.

File: @here/harp-mapview/lib/Utils.ts (hunk 2 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/Utils.ts:187]: yes, fixed

File: @here/harp-mapview/lib/Utils.ts (hunk 3 above)
Comment (atomicsulfate) [reply @here/harp-mapview/lib/Utils.ts]: changed as discussed.

**Pull request merged**